### PR TITLE
Consolidate configmap

### LINKED ISSUE #405: Leverage eventing-kafka field of config-kafka in consolidated channel and in source
**Problem**
There's an `eventing-kafka` [field](https://github.com/knative-sandbox/eventing-kafka/blob/master/config/channel/distributed/300-eventing-kafka-configmap.yaml#L39) in `config-kafka` of the distributed channel:

```yaml
eventing-kafka: |
    receiver:
      cpuRequest: 100m
      memoryRequest: 50Mi
      replicas: 1
    dispatcher:
      cpuRequest: 100m
      memoryRequest: 50Mi
      replicas: 1
    kafka:
      brokers: REPLACE_WITH_CLUSTER_URL
      enableSaramaLogging: false
      topic:
        defaultNumPartitions: 4
        defaultReplicationFactor: 1 # Cannot exceed the number of Kafka Brokers!
        defaultRetentionMillis: 604800000  # 1 week
      adminType: kafka # One of "kafka", "azure", "custom"
```

To be consistent (see the umbrella ticket, https://github.com/knative-sandbox/eventing-kafka/issues/404), we can also leverage these settings in consolidated channel and also in Kafka source.

To talk about fields specifically:
- `receiver`: could only be used in distributed channel
- `dispatcher`: could be used in distributed and consolidated channel
- `kafka.brokers`: Consolidated channel has a separete field in `config-kafka` called `bootstrapServers`. We should leverage that in distributed channel too. `eventing-kafka.kafka.brokers` then would be used as a default OR if possible, we can get rid of that.
- `kafka.enableSaramaLogging`: could be used in all 3 components (2 channels and the source)
- `kafka.topic`: could be also used in consolidated channel
- `kafka.adminType`: could be also used in consolidated channel

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)**
Which persona is this feature for?

**Exit Criteria**
A measurable (binary) test that would indicate that the problem has been resolved.

**Time Estimate (optional):**
How many developer-days do you think this may take to resolve?

**Additional context (optional)**
Add any other context about the feature request here.

### PR DESCRIPTION
**Pull request by eric-sap**
Fixes #405 

## Proposed Changes

- 🧽  Changed sarama.LoadSettings() to return only one "settings" struct (the EventingKafkaConfig) which includes the Sarama config (with auth embedded) as a separate entry, and refactored the related calling functions
- 🧽  Moved a fair bit of sarama-related code into common for use by consolidated and distributed channels
- 🧽  Reorganized config-kafka configmap for better use by multiple components (consolidated and distributed) that don't have the same dispatcher/receiver structure
- 🧽  Made "replicas" an optional entry (default 1) for the dispatcher and receiver config fields with the expectation that HorizontalPodAutoscaler resources will be the preferred mechanism for adjusting these
- 🧽 Moved the "enableSaramaLogging" field from eventing-kafka to the sarama field "enableLogging" (and moved the sarama config string one level down to the "config" field

**Release Note**

```release-note
The current config-kafka configmap format has changed (no immediate action required; the old format will continue to work)
```

### CODE REVIEW

**Comment by eric-sap** [comment]
/hold
I would like to hold this for review for at least a few days as this is a cross-channel-type change and it will be far easier to address any desired adjustments to the configmap now rather than after it's in use.

File: pkg/channel/distributed/controller/testing/data.go
Comment (travis-minke-sap) [inline pkg/channel/distributed/controller/testing/data.go:168]: Is there another real (non-test) constant for this elsewhere that should be used ?

File: test/e2e-common.sh
Comment (travis-minke-sap) [inline test/e2e-common.sh:380]: do we need to delete the receiver Service as well ?

**Review by travis-minke-sap — CHANGES_REQUESTED** [review]
looking good - thanks Eric!

File: pkg/common/config/eventingkafkaconfig.go
Comment (travis-minke-sap) [inline pkg/common/config/eventingkafkaconfig.go:85]: just wondering if the ConfigMap would "read" better (more consistent) if the `Dispatcher` was under Distributed AND Consolidated.  I realize it is "common" to both Channels but a third channel could come along with a different architecture?  It also aligns nicely next to the "Receiver" in distributed case.  Users would only ever have one type of ConfigMap (consolidated / distributed) so it wouldn't appear in there twice.   Anyway - not a big deal, just a minor organziation preference - would be good to hear from others as well ; )

File: pkg/channel/distributed/controller/testing/data.go
Comment (eric-sap) [reply pkg/channel/distributed/controller/testing/data.go:168]: Originally I made this its own constant in the test code specifically so that if we change the "real" version (CurrentConfigVersion in common/config/eventingkafkaconfig.go) without paying attention to the test, then the test will fail.  However ... the upgrade code became less complicated and the tests that used it aren't there, so now it's only in TestNewController(), which should be using the common constant.  Will fix.

File: pkg/common/kafka/sarama/sarama.go
Comment (eric-sap) [reply pkg/common/kafka/sarama/sarama.go:49]: Yep; this used to be a test-only function and when I moved it out to real-code use I neglected to pay attention to that; will fix.

File: test/e2e-common.sh
Comment (eric-sap) [reply test/e2e-common.sh:380]: That would be a good idea; I forgot about that.  Distributed is the last test before the cluster gets torn down anyway but it's good to be clean.

File: config/channel/distributed/300-eventing-kafka-configmap.yaml
Comment (eric-sap) [reply config/channel/distributed/300-eventing-kafka-configmap.yaml:54]: This shouldn't be much trouble if desired; a change to consolidated/reconciler/controller/kafkachannel.go where it calls kafkaClusterAdmin.CreateTopic()  (that function now has access to r.kafkaConfig.EventingKafka.Kafka.Topic so it seems easy enough)

Comment (eric-sap) [reply config/channel/distributed/300-eventing-kafka-configmap.yaml:54]: I added this (and moved the NumPartitions/etc. functions to common/config/util.go)

File: pkg/common/config/eventingkafkaconfig.go
Comment (eric-sap) [reply pkg/common/config/eventingkafkaconfig.go:87]: I tried it out and it does look less wonky that way.  New configmap format:
```
  eventing-kafka: |
    ...
    channel:
      distributed:
        adminType: kafka
        dispatcher:
          cpuRequest: 100m
          memoryRequest: 50Mi
        receiver:
          cpuRequest: 100m
          memoryRequest: 50Mi
```
or
```
  eventing-kafka: |
    ...
    channel:
      consolidated:
        dispatcher:
          cpuRequest: 100m
          memoryRequest: 50Mi
```
And as an extra bonus, the EKConsolidatedConfig struct is no longer "reserved for future use"

File: pkg/channel/consolidated/utils/util.go
Comment (travis-minke-sap) [inline pkg/channel/consolidated/utils/util.go:118]: This seems odd - we're adding an `error` return value but then ALWAYS returning nil ?  What value does it provide?

Comment (eric-sap) [reply pkg/channel/consolidated/utils/util.go:118]: It's allowing the GetKafkaAuthData function to satisfy the sarama.GetAuth prototype so it can be passed in directly to the LoadSettings call in GetKafkaConfig().  Otherwise it would need a one-line wrapper to return the nil error (which I would be happy to do if that would look better).

File: pkg/channel/distributed/controller/config/config_test.go
Comment (eric-sap) [reply pkg/channel/distributed/controller/config/config_test.go:49]: You're right; I was originally going to call VerifyConfiguration in consolidated as well, but decided that was not the scope here and forgot to remove this kind of stuff.  I'll clean it up.

File: pkg/common/config/util.go
Comment (travis-minke-sap) [inline pkg/common/config/util.go:86]: The `eventing-contrib` comment is out-of-date and should be reworded to reflect the history.  Essentially the current KafkaChannel spec doesn't have this and maybe we should look at adding it (was in old versions of distributed channel CRD.)

File: pkg/common/config/util_test.go
Comment (travis-minke-sap) [inline pkg/common/config/util_test.go:189]: same - fix comment

**Comment by travis-minke-sap** [comment]
I added a couple more trivial comments but otherwise LGTM - will wait for a "consolidated" expert to do final approval.
Lots of great work here - thanks Eric!

File: pkg/common/config/util.go
Comment (eric-sap) [reply pkg/common/config/util.go:86]: I removed the RetentionMillis call and am now just using the Kafka.Topic.DefaultRetentionMillis value from the configmap directly where it is needed (consolidated reconcileTopic and distributed reconcileKafkaTopic).  I put the updated TODO comment in those locations so it will not be lost.

File: pkg/common/config/util_test.go
Comment (eric-sap) [reply pkg/common/config/util_test.go:189]: Functionality/Comment has been moved (see previous review comment)

**Comment by eric-sap** [comment]
/retest
The linter appears to have .. crashed?

**Comment by eric-sap** [comment]
The test are passing and I think this is ready but I'd like to leave it for just a bit more in a final state in case someone from the consolidated side wants to weigh in.

**Comment by eric-sap** [comment]
This is showing up in the kafka-go-coverage test, and I'm fairly certain it's the extra newline in the Bearer field (which is why the length is 41), but I don't know what's creating that value.
```
2021/05/21 01:30:22 Running workflow: presubmit
Warning: token len = 41, which is different from 40
...
2021/05/21 01:30:22 error running c.PullRequests.ListFiles(Ctx, repoOwner=knative-sandbox, RepoName=eventing-kafka, pullNum=622, listOptions): Get "https://api.github.com/repos/knative-sandbox/eventing-kafka/pulls/622/files?page=1": net/http: invalid header field value "Bearer ghp_tYww[...]7wKf\n" for key Authorization
```

**Comment by eric-sap** [comment]
@chizhg  much appreciated!

File: config/channel/distributed/300-eventing-kafka-configmap.yaml
Comment (aliok) [inline config/channel/distributed/300-eventing-kafka-configmap.yaml:55]: I am not 100% sure but I think it might be a better idea to not have the separation of distributed/consolidated in the configmap. In the long run, if we want to have a single channel, users would do another migration of the configmap.

There's pros/cons in both. I would go for not having `consolidated` or `distributed` nested fields and let the implementations ignore the fields they don't use. Just a thought but I am also fine with the approach in the PR.

Comment (aliok) [reply config/channel/distributed/300-eventing-kafka-configmap.yaml:55]: In the end, users can't use both channels at the same time anyway.

File: pkg/common/client/admin_client.go
Comment (aliok) [inline pkg/common/client/admin_client.go:9]: Maybe we can get rid of this function and the file?

File: test/config/config-kafka-sasl.yaml
Comment (aliok) [inline test/config/config-kafka-sasl.yaml:6]:
```suggestion
  # eventing-kafka.kafka.brokers: Replace this with the URLs for your kafka cluster,
```

File: test/config/config-kafka-tls.yaml
Comment (aliok) [inline test/config/config-kafka-tls.yaml:6]:
```suggestion
  # eventing-kafka.kafka.brokers: Replace this with the URLs for your kafka cluster,
```

File: test/e2e-common.sh
Comment (aliok) [inline test/e2e-common.sh:381]: `32603413` is there a deterministic process to generate this number or is this just forgotten?

**Comment by aliok** [comment]
@eric-sap I added a few comments. Let's resolve those and I can have a 2nd review.

Thanks a lot for the PR! I truly appreciate your ~30 commits and patience with this consolidation PR. Awesome work!

File: test/e2e-common.sh
Comment (eric-sap) [reply test/e2e-common.sh:381]: Yes, this comes from GenerateHash() in pkg/channel/distributed/controller/util/hash.go
For a secret named "kafka-cluster" the hash will always be 32603413
We could replicate the GenerateHash function in shell script here if that's more understandable; it would look something like this but with a parameter
```
echo -n "kafka-cluster" | md5 | sed 's/^\(........\).*/\1/g'
```

File: test/config/config-kafka-sasl.yaml
Comment (eric-sap) [reply test/config/config-kafka-sasl.yaml:6]: got it, thanks

File: test/config/config-kafka-tls.yaml
Comment (eric-sap) [reply test/config/config-kafka-tls.yaml:6]: got it, thanks

File: pkg/common/client/admin_client.go
Comment (eric-sap) [reply pkg/common/client/admin_client.go:9]: Good idea; will remove

File: test/e2e-common.sh
Comment (eric-sap) [reply test/e2e-common.sh:381]: We are planning, however (with issue #608), to use a non-hash-decorated single-name for the receiver, so I don't think adding this hash function (simple though it is) to the bash script will really be a long-term gain in usefulness.  (It does need to stay in the go code; the same function is used for the dispatcher names -- those pods are removed automatically when the kafkachannel is deleted during the test so a similar kubectl command isn't needed here)

Comment (eric-sap) [reply test/e2e-common.sh]: This ended up being such a small change that I opted for the visibility of adding the hash generation and a comment about the reason for the value, even if it ends up being temporary.

File: config/channel/consolidated/configmaps/kafka-config.yaml
Comment (matzew) [inline config/channel/consolidated/configmaps/kafka-config.yaml:21]: @eric-sap can we have some doc update for these changes on the docs repo:
https://github.com/knative/docs/tree/mkdocs/docs/eventing/samples/kafka/channel#authentication-against-an-apache-kafka

File: pkg/common/kafka/sarama/sarama.go
Comment (matzew) [inline pkg/common/kafka/sarama/sarama.go:266]: can both call into the same "helper" ?

**Comment by matzew** [comment]
LGTM

File: test/e2e-common.sh
Comment (aliok) [reply test/e2e-common.sh]: Thanks. I am happy with both as long as there's determinism :)

**Review by aliok — APPROVED** [review]
/lgtm
/approve

Thanks for addressing the comments.

**Comment by aliok** [comment]
@eric-sap I didn't want to unhold your `/hold` ;)

File: pkg/common/kafka/sarama/sarama.go
Comment (eric-sap) [reply pkg/common/kafka/sarama/sarama.go:266]: I have tried to combine these functions and the result is not any better than these separate ones because of the amount of overhead necessary ("slice of pointers" is different than "slice of RecordHeaders", despite go papering over that and allowing you to write "header.Value" for either).  A Go master might know something better (happy to accept something!) but I threw in the towel.

File: config/channel/consolidated/configmaps/kafka-config.yaml
Comment (eric-sap) [reply config/channel/consolidated/configmaps/kafka-config.yaml:21]: Absolutely; I will get on that.  Thank you for the pointer.

**Pull request merged**
